src/groth16/aggregate/commit.rs

```diff
@@ -0,0 +1,152 @@
+    /// In other words, it returns $\{v_i^{s_i}\}$
+    pub fn scale(&self, s_vec: &[G::Scalar]) -> Self {
+        let (a, b) = self
+            .a
+            .par_iter()
+            .zip(self.b.par_iter())
+            .zip(s_vec.par_iter())
+            .map(|((ap, bp), si)| {
+                let v1s = mul!(ap.into_projective(), si.into_repr()).into_affine();
+                let v2s = mul!(bp.into_projective(), si.into_repr()).into_affine();
+                (v1s, v2s)
+            })
+            .unzip();
+
+        Self { a: a, b: b }
+    }
+
+    /// split returns the left and right commitment key part. It makes copy.
+    /// TODO: remove the copy
+    pub fn split(&self, at: usize) -> (Self, Self) {
+        let (a_l, a_r) = self.a.split_at(at);
+        let (b_l, b_r) = self.b.split_at(at);
+        (
+            Self {
+                a: a_l.to_vec(),
+                b: b_l.to_vec(),
+            },
+            Self {
+                a: a_r.to_vec(),
+                b: b_r.to_vec(),
+            },
+        )
+    }
+
+    /// Compress takes a left and right commitment key and returns a commitment
+    /// key $left \circ right^{scale} = (left_i*right_i^{scale} ...)$. This is
+    /// required step during GIPA recursion.
+    pub fn compress(left: &Self, right: &Self, scale: &G::Scalar) -> Self {
+        assert!(left.a.len() == right.a.len());
+        let (a, b): (Vec<G>, Vec<G>) = left
+            .a
+            .par_iter()
+            .zip(left.b.par_iter())
+            .zip(right.a.par_iter())
+            .zip(right.b.par_iter())
+            .map(|(((left_a, left_b), right_a), right_b)| {
+                let mut ra = mul!(right_a.into_projective(), scale.into_repr());
+                let mut rb = mul!(right_b.into_projective(), scale.into_repr());
+                ra.add_assign_mixed(left_a);
+                rb.add_assign_mixed(left_b);
+                (ra.into_affine(), rb.into_affine())
+            })
+            .unzip();
+
+        assert!(a.len() == left.a.len());
+        assert!(b.len() == left.a.len());
+        Self { a: a, b: b }
+    }
+
+    /// first returns the first values in the vector of v1 and v2 (respectively
+    /// w1 and w2). When commitment key is of size one, it's a proxy to get the
+    /// final values.
+    pub fn first(&self) -> (G, G) {
+        assert!(self.a.len() == 1 && self.b.len() == 1);
+        (self.a[0].clone(), self.b[0].clone())
+    }
+}
+
+/// Both commitment outputs a pair of $F_q^k$ element.
+pub type Output<E> = (<E as Engine>::Fqk, <E as Engine>::Fqk);
+
+/// single_g1 commits to a single vector of G1 elements in the following way:
+/// $T = \prod_{i=0}^n e(A_i, v_{1,i})$
+/// $U = \prod_{i=0}^n e(A_i, v_{2,i})$
+/// Output is $(T,U)$
+pub fn single_g1<E: Engine>(vkey: &VKey<E>, a: &[E::G1Affine]) -> Output<E> {
+    let t = inner_product::pairing::<E>(a, &vkey.a);
+    let u = inner_product::pairing::<E>(a, &vkey.b);
+    return (t, u);
```

Comment:
no `return` at the end

Reply:
I'll remove it but not without saying it is still valid rust to write return !!